File: frontend/src/api/postgresql.ts
import request, { getToken, type PageResult } from './request'

export interface PostgreSQLInstanceInfo {
  config: {
    host: string
    port: string
    database: string
    user: string
    password: string
  }
  status: string
  message: string
  version: string
  currentDb: string
  currentUser: string
  role: string
  serverAddr: string
  serverPort: number
  settings: Record<string, string>
  replications: {
    clientAddr: string
    state: string
    syncState: string
    writeLag: string
    flushLag: string
    replayLag: string
  }[]
  checkedAt: string
}

export interface SQLChangeFile {
  id: number
  batchId: number
  batchSortNo: number
  systemCode: string
  environment: string
  schemaName: string
  version: string
  groupSortNo: number
  fileName: string
  fileContent?: string
  executeStatus: string
  executeMessage: string
  executeUser: string
  executeTime?: string
  createdAt: string
}

export interface SQLChangeBatch {
  id: number
  batchName: string
  executeStatus: string
  executeMessage: string
  executeUser: string
  executeTime?: string
  totalFiles: number
  successFiles: number
  failedFiles: number
  skippedFiles: number
  createdAt: string
}

export interface SQLChangeStatement {
  id: number
  fileId: number
  lineNumber: number
  sqlContent: string
  sqlType: string
  riskLevel: string
  riskReason: string
  canRunInTransaction: boolean
  executionStrategy: string
  executeStatus: string
  executeMessage: string
  sqlState: string
  estimatedRows: number
  affectedRows: number
  durationMs: number
  executeTime?: string
}

export interface ParseSQLRequest {
  systemCode?: string
  environment?: string
  schemaName?: string
  version?: string
  groupSortNo?: number
  fileName?: string
  content: string
  overwrite?: boolean
}

export interface SQLExecuteOptions {
  skipExistsColumn: boolean
  skipExistsTable: boolean
  skipUniqueConstraint: boolean
  requireRiskConfirmation: boolean
  confirmWarnRisk: boolean
  forceBlockedSql: boolean
}

export interface SQLForceWhitelistOption {
  sqlType: string
  label: string
}

export interface SQLForceWhitelistConfig {
  available: SQLForceWhitelistOption[]
  enabled: string[]
}

export interface ParseSQLBatchFile {
  fileName: string
  content: string
}

export interface SQLViewDependencyTask {
  id: number
  schemaName: string
  tableName: string
  columnName: string
  alterSql: string
  status: string
  riskLevel: string
  riskReason: string
  lockTimeout: string
  statementTimeout: string
  operator: string
  executeMessage: string
  analyzedAt?: string
  executedAt?: string
  createdAt: string
}

export interface SQLViewDependencyItem {
  id: number
  taskId: number
  objectSchema: string
  objectName: string
  objectKind: string
  depth: number
  dropOrder: number
  restoreOrder: number
  definition: string
  ownerName: string
  grantsJson: string
  commentsJson: string
  indexesJson: string
  rulesJson: string
  triggersJson: string
  optionsJson: string
  backupHash: string
  dropSql: string
  createSql: string
  restoreRefreshSql: string
  restoreOwnerSql: string
  restoreGrantsSql: string
  restoreCommentsSql: string
  restoreIndexesSql: string
  restoreRulesSql: string
  restoreTriggersSql: string
  verifySql: string
  status: string
  errorMessage: string
}

export interface SQLViewDependencyTaskRequest {
  schemaName: string
  tableName: string
  columnName: string
  alterSql: string
  lockTimeout?: string
  statementTimeout?: string
}

export const getPostgreSQLInstance = () =>
  request.get<any, PostgreSQLInstanceInfo>('/postgresql/instance')

export const getSQLForceWhitelist = () =>
  request.get<any, SQLForceWhitelistConfig>('/postgresql/sql-force-whitelist')

export const saveSQLForceWhitelist = (enabled: string[]) =>
  request.put<any, SQLForceWhitelistConfig>('/postgresql/sql-force-whitelist', { enabled })

export const listSQLViewDependencyTasks = (page = 1, pageSize = 20) =>
  request.get<any, PageResult<SQLViewDependencyTask>>('/postgresql/view-dependency-tasks', { params: { page, pageSize } })

export const createSQLViewDependencyTask = (data: SQLViewDependencyTaskRequest) =>
  request.post<any, SQLViewDependencyTask>('/postgresql/view-dependency-tasks', data)

export const getSQLViewDependencyTask = (id: number) =>
  request.get<any, { task: SQLViewDependencyTask; items: SQLViewDependencyItem[] }>(`/postgresql/view-dependency-tasks/${id}`)

export const analyzeSQLViewDependencyTask = (id: number) =>
  request.post<any, { task: SQLViewDependencyTask; items: SQLViewDependencyItem[] }>(`/postgresql/view-dependency-tasks/${id}/analyze`)

export const precheckSQLViewDependencyTask = (id: number) =>
  request.post<any, { task: SQLViewDependencyTask; items: SQLViewDependencyItem[] }>(`/postgresql/view-dependency-tasks/${id}/precheck`)

export const executeSQLViewDependencyTask = (id: number) =>
  request.post<any, { task: SQLViewDependencyTask; items: SQLViewDependencyItem[] }>(`/postgresql/view-dependency-tasks/${id}/execute`)

export const restoreSQLViewDependencyTask = (id: number) =>
  request.post<any, { task: SQLViewDependencyTask; items: SQLViewDependencyItem[] }>(`/postgresql/view-dependency-tasks/${id}/restore`)

export const parseSQLFile = (data: ParseSQLRequest) =>
  request.post<any, { file: SQLChangeFile; statements: SQLChangeStatement[] }>('/postgresql/sql-files/parse', data)

export const executeSQLFile = (id: number) =>
  request.post<any, { file: SQLChangeFile; statements: SQLChangeStatement[] }>(`/postgresql/sql-files/${id}/execute`)

export const executeSQLFileWithOptions = (id: number, options: SQLExecuteOptions) =>
  request.post<any, { file: SQLChangeFile; statements: SQLChangeStatement[] }>(`/postgresql/sql-files/${id}/execute`, { options })

export const executeSQLContent = (data: ParseSQLRequest & { options: SQLExecuteOptions }) =>
  request.post<any, { file: SQLChangeFile; statements: SQLChangeStatement[] }>('/postgresql/sql-files/execute-content', data)

export const cancelSQLFile = (id: number) =>
  request.post<any, SQLChangeFile>(`/postgresql/sql-files/${id}/cancel`)

export const parseSQLBatch = (data: { batchName?: string; overwrite?: boolean; files: ParseSQLBatchFile[] }) =>
  request.post<any, { batch: SQLChangeBatch; files: SQLChangeFile[] }>('/postgresql/sql-batches/parse', data)

export const executeSQLBatch = (id: number, options: SQLExecuteOptions) =>
  request.post<any, { batch: SQLChangeBatch; files: SQLChangeFile[] }>(`/postgresql/sql-batches/${id}/execute`, { options })

export const cancelSQLBatch = (id: number) =>
  request.post<any, SQLChangeBatch>(`/postgresql/sql-batches/${id}/cancel`)

export const listSQLBatches = (page = 1, pageSize = 20) =>
  request.get<any, PageResult<SQLChangeBatch>>('/postgresql/sql-batches', { params: { page, pageSize } })

export const getSQLBatch = (id: number) =>
  request.get<any, { batch: SQLChangeBatch; files: SQLChangeFile[] }>(`/postgresql/sql-batches/${id}`)

export const listSQLFiles = (page = 1, pageSize = 20) =>
  request.get<any, PageResult<SQLChangeFile>>('/postgresql/sql-files', { params: { page, pageSize } })

export const listTodoSQLFiles = (page = 1, pageSize = 20) =>
  request.get<any, PageResult<SQLChangeFile>>('/postgresql/sql-files/todo', { params: { page, pageSize } })

export const listDoneSQLFiles = (page = 1, pageSize = 20) =>
  request.get<any, PageResult<SQLChangeFile>>('/postgresql/sql-files/done', { params: { page, pageSize } })

export const getSQLFile = (id: number) =>
  request.get<any, { file: SQLChangeFile; statements: SQLChangeStatement[] }>(`/postgresql/sql-files/${id}`)

export const skipSQLStatement = (id: number) =>
  request.post<any, SQLChangeStatement>(`/postgresql/sql-statements/${id}/skip`)

export const skipSQLFile = (id: number) =>
  request.post<any, SQLChangeFile>(`/postgresql/sql-files/${id}/skip`)

export const deleteSQLFile = (id: number) =>
  request.delete<any, { deleted: boolean }>(`/postgresql/sql-files/${id}`)

export const importServerSQL = (filePath: string, overwrite = true) =>
  request.post<any, { count: number }>('/postgresql/sql-files/import-server', { filePath, overwrite })

export const exportNotExecutableSQLUrl = (id: number) =>
  `/api/postgresql/sql-files/${id}/not-executable.sql`

export const exportSQLViewDependencyPlanUrl = (id: number) =>
  `/api/postgresql/view-dependency-tasks/${id}/plan.sql`

export const exportSQLViewDependencyRestorePlanUrl = (id: number) =>
  `/api/postgresql/view-dependency-tasks/${id}/restore.sql`

export async function exportSQLViewDependencyRestorePlan(id: number) {
  const res = await fetch(exportSQLViewDependencyRestorePlanUrl(id), {
    headers: { Authorization: `Bearer ${getToken()}` },
  })
  if (!res.ok) throw new Error('导出失败')
  return await res.text()
}

export async function exportSQLViewDependencyPlan(id: number) {
  const res = await fetch(exportSQLViewDependencyPlanUrl(id), {
    headers: { Authorization: `Bearer ${getToken()}` },
  })
  if (!res.ok) throw new Error('导出失败')
  return await res.text()
}

export async function exportNotExecutableSQL(id: number) {
  const res = await fetch(exportNotExecutableSQLUrl(id), {
    headers: { Authorization: `Bearer ${getToken()}` },
  })
  if (!res.ok) throw new Error('导出失败')
  return await res.text()
}

File: backend/internal/model/postgresql.go
package model

import "time"

type SQLChangeFile struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	BatchID        uint       `gorm:"index" json:"batchId"`
	BatchSortNo    int        `json:"batchSortNo"`
	SystemCode     string     `gorm:"size:64" json:"systemCode"`
	Environment    string     `gorm:"size:64" json:"environment"`
	SchemaName     string     `gorm:"size:64" json:"schemaName"`
	Version        string     `gorm:"size:64" json:"version"`
	GroupSortNo    int        `json:"groupSortNo"`
	FileName       string     `gorm:"size:255" json:"fileName"`
	FileContent    string     `gorm:"type:text" json:"fileContent"`
	ExecuteStatus  string     `gorm:"size:24;index" json:"executeStatus"`
	ExecuteMessage string     `gorm:"type:text" json:"executeMessage"`
	ExecuteUser    string     `gorm:"size:64" json:"executeUser"`
	ExecuteTime    *time.Time `json:"executeTime"`
	IsDeleted      bool       `gorm:"index" json:"isDeleted"`
	CreatedAt      time.Time  `json:"createdAt"`
	UpdatedAt      time.Time  `json:"updatedAt"`
}

type SQLChangeBatch struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	BatchName      string     `gorm:"size:255" json:"batchName"`
	ExecuteStatus  string     `gorm:"size:24;index" json:"executeStatus"`
	ExecuteMessage string     `gorm:"type:text" json:"executeMessage"`
	ExecuteUser    string     `gorm:"size:64" json:"executeUser"`
	ExecuteTime    *time.Time `json:"executeTime"`
	TotalFiles     int        `json:"totalFiles"`
	SuccessFiles   int        `json:"successFiles"`
	FailedFiles    int        `json:"failedFiles"`
	SkippedFiles   int        `json:"skippedFiles"`
	IsDeleted      bool       `gorm:"index" json:"isDeleted"`
	CreatedAt      time.Time  `json:"createdAt"`
	UpdatedAt      time.Time  `json:"updatedAt"`
}

type SQLChangeStatement struct {
	ID                  uint       `gorm:"primaryKey" json:"id"`
	FileID              uint       `gorm:"index" json:"fileId"`
	LineNumber          int        `json:"lineNumber"`
	SQLContent          string     `gorm:"type:text" json:"sqlContent"`
	SQLType             string     `gorm:"size:64" json:"sqlType"`
	RiskLevel           string     `gorm:"size:24" json:"riskLevel"`
	RiskReason          string     `gorm:"type:text" json:"riskReason"`
	EstimatedRows       int64      `json:"estimatedRows"`
	CanRunInTransaction bool       `json:"canRunInTransaction"`
	ExecutionStrategy   string     `gorm:"size:32" json:"executionStrategy"`
	ExecuteStatus       string     `gorm:"size:24;index" json:"executeStatus"`
	ExecuteMessage      string     `gorm:"type:text" json:"executeMessage"`
	SQLState            string     `gorm:"size:16" json:"sqlState"`
	AffectedRows        int64      `json:"affectedRows"`
	DurationMs          int64      `json:"durationMs"`
	ExecuteTime         *time.Time `json:"executeTime"`
	CreatedAt           time.Time  `json:"createdAt"`
	UpdatedAt           time.Time  `json:"updatedAt"`
}

type SQLViewBackup struct {
	ID           uint      `gorm:"primaryKey" json:"id"`
	FileID       uint      `gorm:"index" json:"fileId"`
	StatementID  uint      `gorm:"index" json:"statementId"`
	SchemaName   string    `gorm:"size:64;index" json:"schemaName"`
	ViewName     string    `gorm:"size:128;index" json:"viewName"`
	Definition   string    `gorm:"type:text" json:"definition"`
	DropSQL      string    `gorm:"type:text" json:"dropSql"`
	CreateSQL    string    `gorm:"type:text" json:"createSql"`
	BackupReason string    `gorm:"type:text" json:"backupReason"`
	CreatedAt    time.Time `json:"createdAt"`
}

type SQLViewDependencyTask struct {
	ID               uint       `gorm:"primaryKey" json:"id"`
	SchemaName       string     `gorm:"size:64;index" json:"schemaName"`
	TableName        string     `gorm:"size:128;index" json:"tableName"`
	ColumnName       string     `gorm:"size:128;index" json:"columnName"`
	AlterSQL         string     `gorm:"type:text" json:"alterSql"`
	Status           string     `gorm:"size:32;index" json:"status"`
	RiskLevel        string     `gorm:"size:24" json:"riskLevel"`
	RiskReason       string     `gorm:"type:text" json:"riskReason"`
	LockTimeout      string     `gorm:"size:32" json:"lockTimeout"`
	StatementTimeout string     `gorm:"size:32" json:"statementTimeout"`
	Operator         string     `gorm:"size:64" json:"operator"`
	ExecuteMessage   string     `gorm:"type:text" json:"executeMessage"`
	AnalyzedAt       *time.Time `json:"analyzedAt"`
	ExecutedAt       *time.Time `json:"executedAt"`
	IsDeleted        bool       `gorm:"index" json:"isDeleted"`
	CreatedAt        time.Time  `json:"createdAt"`
	UpdatedAt        time.Time  `json:"updatedAt"`
}

type SQLViewDependencyItem struct {
	ID                 uint      `gorm:"primaryKey" json:"id"`
	TaskID             uint      `gorm:"index" json:"taskId"`
	ObjectSchema       string    `gorm:"size:64;index" json:"objectSchema"`
	ObjectName         string    `gorm:"size:128;index" json:"objectName"`
	ObjectKind         string    `gorm:"size:8" json:"objectKind"`
	Depth              int       `json:"depth"`
	DropOrder          int       `json:"dropOrder"`
	RestoreOrder       int       `json:"restoreOrder"`
	Definition         string    `gorm:"type:text" json:"definition"`
	OwnerName          string    `gorm:"size:128" json:"ownerName"`
	GrantsJSON         string    `gorm:"type:text" json:"grantsJson"`
	CommentsJSON       string    `gorm:"type:text" json:"commentsJson"`
	IndexesJSON        string    `gorm:"type:text" json:"indexesJson"`
	RulesJSON          string    `gorm:"type:text" json:"rulesJson"`
	TriggersJSON       string    `gorm:"type:text" json:"triggersJson"`
	OptionsJSON        string    `gorm:"type:text" json:"optionsJson"`
	BackupHash         string    `gorm:"size:128;index" json:"backupHash"`
	DropSQL            string    `gorm:"type:text" json:"dropSql"`
	CreateSQL          string    `gorm:"type:text" json:"createSql"`
	RestoreRefreshSQL  string    `gorm:"type:text" json:"restoreRefreshSql"`
	RestoreOwnerSQL    string    `gorm:"type:text" json:"restoreOwnerSql"`
	RestoreGrantsSQL   string    `gorm:"type:text" json:"restoreGrantsSql"`
	RestoreCommentsSQL string    `gorm:"type:text" json:"restoreCommentsSql"`
	RestoreIndexesSQL  string    `gorm:"type:text" json:"restoreIndexesSql"`
	RestoreRulesSQL    string    `gorm:"type:text" json:"restoreRulesSql"`
	RestoreTriggersSQL string    `gorm:"type:text" json:"restoreTriggersSql"`
	VerifySQL          string    `gorm:"type:text" json:"verifySql"`
	Status             string    `gorm:"size:32;index" json:"status"`
	ErrorMessage       string    `gorm:"type:text" json:"errorMessage"`
	CreatedAt          time.Time `json:"createdAt"`
	UpdatedAt          time.Time `json:"updatedAt"`
}

type SQLViewDependencyStep struct {
	ID           uint       `gorm:"primaryKey" json:"id"`
	TaskID       uint       `gorm:"index" json:"taskId"`
	StepName     string     `gorm:"size:64;index" json:"stepName"`
	StepOrder    int        `json:"stepOrder"`
	SQLContent   string     `gorm:"type:text" json:"sqlContent"`
	Status       string     `gorm:"size:32;index" json:"status"`
	ErrorMessage string     `gorm:"type:text" json:"errorMessage"`
	StartedAt    *time.Time `json:"startedAt"`
	FinishedAt   *time.Time `json:"finishedAt"`
	CreatedAt    time.Time  `json:"createdAt"`
	UpdatedAt    time.Time  `json:"updatedAt"`
}
